demo/GraphTutorial/app/src/main/java/com/example/graphtutorial/GraphHelper.java

```diff
@@ -90,34 +72,49 @@ public void getCalendarView(String accessToken,
         final List<Option> pagingOptions = new LinkedList<Option>();
         pagingOptions.add(new HeaderOption("Prefer", "outlook.timezone=\"" + timeZone + "\""));
 
-        final ICallback<IEventCollectionPage> pagingCallback = new ICallback<IEventCollectionPage>() {
-            @Override
-            public void success(IEventCollectionPage eventCollectionPage) {
-                allEvents.addAll(eventCollectionPage.getCurrentPage());
-
-                IEventCollectionRequestBuilder nextPage =
-                        eventCollectionPage.getNextPage();
-
-                if (nextPage == null) {
-                    callback.success(allEvents);
-                }
-                else{
-                    nextPage.buildRequest(pagingOptions)
-                            .get(this);
-                }
-            }
-
-            @Override
-            public void failure(ClientException ex) {
-                callback.failure(ex);
-            }
-        };
+        final CompletableFuture<List<Event>> future = new CompletableFuture<>();
+
+        mClient.me().calendarView()
+            .buildRequest(options)
+            .select("subject,organizer,start,end")
+            .orderBy("start/dateTime")
+            .top(25)
+            .getAsync()
+            .thenAccept(eventPage -> {
+                processPage(eventPage, allEvents, pagingOptions, future);
+            })
+            .exceptionally(exception -> {
+                future.completeExceptionally(exception);
+                return null;
+            });
+
+        return future;
+    }
 
-        mClient.me().calendarView().buildRequest(options)
-                .select("subject,organizer,start,end")
-                .orderBy("start/dateTime")
-                .top(25)
-                .get(pagingCallback);
+    private void processPage(EventCollectionPage currentPage,
+                             List<Event> eventList,
+                             List<Option> options,
+                             CompletableFuture<List<Event>> future) {
+        eventList.addAll(currentPage.getCurrentPage());
+
+        // Check if there is another page of results
+        EventCollectionRequestBuilder nextPage = currentPage.getNextPage();
+        if (nextPage != null) {
+            // Request the next page and repeat
+            nextPage.buildRequest(options)
```

Comment:
you probably need to block here or call composeWith to avoid running into race conditions.

Reply:
yeah I'm not familiar enough to foresee the behavior here. My guess is if you rendered that collection to some kind of table view and if that table view was sorted by a field, things would flash a couple of times while it's loading. (or you'd have to sent the sort order to the service)